File: 0779-k-th-symbol-in-grammar/0779-k-th-symbol-in-grammar.py
class Solution:
    def kthGrammar(self, n: int, k: int) -> int:
        if n==1 and k==1:
            return 0
        temp=1
        # if k%2==0:
        #     temp=self.kthGrammar(n-1,k//2)
        # else:
        #     temp=self.kthGrammar(n-1,(k//2)+1)
        
        if k<=(2**(n-1)//2):
            temp=self.kthGrammar(n-1,k)
        else:
            temp=0 if self.kthGrammar(n-1,k-(2**(n-1)//2))==1 else 1
            
        # if temp==0:
        #     if k%2==1:
        #         return 0
        #     else:
        #         return 1
        # else:
        #     if k%2==1:
        #         return 1
        #     else:
        #         return 0
        
        return temp

File: 0779-k-th-symbol-in-grammar/0779-k-th-symbol-in-grammar.java
class Solution {
    public static int solve(int n,int k){
        if(n==1 && k==1){
            return 0;
        }
        int t=Solution.solve(n-1,(int)Math.ceil(k/2.0));
        if(t==0){
            if(k%2==0){
                return 1;
            }else{
                return 0;
            }
        }else{
            if(k%2==1){
                return 1;
            }else{
                return 0;
            }
        }
    }
    public int kthGrammar(int n, int k) {
        return Solution.solve(n,k);
        
    }
}